docs: add "Using Autocomplete with Vue/React" guides (#442)
algolia/autocomplete.js: packages/website/docs/using-react.md

---
id: using-react
title: Using Autocomplete with React
---

> Learn how to create and use an React Autocomplete component.

Reviewer: ```suggestion
Learn how to create and use a React Autocomplete component.
```

This guide shows how to create a React Autocomplete component. It uses the [`useRef`](https://reactjs.org/docs/hooks-reference.html#useref) and [`useEffect`](https://reactjs.org/docs/hooks-reference.html#useeffect) hooks to create and mount the component. It doesn't define specific [sources](sources). Rather, you can pass [sources](sources) and other [options](basic-options) as [props](https://reactjs.org/docs/components-and-props.html).

## Prerequisites

This tutorial assumes that you have:

> - an existing [React (v 16.8+) application](https://reactjs.org/docs/getting-started.html) where you want to implement the autocomplete dropdown

Reviewer: ```suggestion
- an existing [React (v16.8+) application](https://reactjs.org/docs/getting-started.html) where you want to implement the autocomplete dropdown
```

- familiarity with the [basic Autocomplete configuration options](basic-options)

## Creating the component

Start with some boilerplate for creating a React component. This component uses the [`useRef`](https://reactjs.org/docs/hooks-reference.html#useref) hook to create a mutable ref object, `containerRef`, to mount the autocomplete on. To learn more about this hook, check out the [`useRef` React documentation](https://reactjs.org/docs/hooks-reference.html#useref).

 All that you need to render is a `div` with the `containerRef` as the `ref`.

```jsx title="Autocomplete.jsx"
import React, { createElement, Fragment, useEffect, useRef } from 'react';
import { render } from 'react-dom';

export function Autocomplete(props) {
  const containerRef = useRef(null);

  useEffect(() => {
    if (!containerRef.current) {
      return undefined;
    }
  }, [props]);

  return (
    <div
      ref={containerRef}
    />
  );
}
```

### Mounting the autocomplete

Now that you have access to the DOM through the `containerRef` object, you can create and mount the autocomplete instance.

> You can use the same [Autocomplete options](basic-options), but you must also pass the `renderer` and `render` parameters. You can rely on `props` to pass any autocomplete options you want to remain configurable.

Reviewer: Wdyt of just adding a sentence that explains why?

```suggestion
You can use the same [Autocomplete options](basic-options), but you must also pass the `renderer` and `render` parameters. Those allow to plug the internal React mechanism into Autocomplete. You can rely on `props` to pass any autocomplete options you want to remain configurable.
```

Author: Good idea! Could you help explain a bit more about what the "React mechanism" is? I think this language is a bit vague, so it might not help the reader understand more.

Reviewer: The default Autocomplete implementation is to use Preact's version of [`createElement`](https://preactjs.com/guide/v10/api-reference/#h--createelement), [`Fragment`](https://preactjs.com/guide/v10/api-reference/#fragment) and [`render`](https://preactjs.com/guide/v10/api-reference/#render).

Each framework has its own specifities for these APIs, and they work together as a whole (we can't mix them). Although the concepts are the same, the expected input and format differ slightly.

This is why we need to provide the React version of [`createElement`](https://reactjs.org/docs/react-api.html#createelement), [`Fragment`](https://reactjs.org/docs/react-api.html#reactfragment) and [`render`](https://reactjs.org/docs/rendering-elements.html#rendering-an-element-into-the-dom).

Once these values are specified, Autocomplete will use them to render the views.

Author: How does this sound:
Maybe it doesn't need to be more exact. How does this sound:

```suggestion
You can use the same [Autocomplete options](basic-options), but you must also pass the `renderer` and `render` parameters. These parameters let the component plug into internal React mechanisms. You can rely on `props` to pass any autocomplete options you want to remain configurable.
```

Author: Oops, missed the previous reply ☝️

Author: https://github.com/algolia/autocomplete.js/pull/442/commits/1332818d65852fb4fa9ea5d273b592cca842ab6c

Don't forget to [clean up the effect](https://reactjs.org/docs/hooks-reference.html#cleaning-up-an-effect) by returning a function that destroys the autocomplete instance.

```jsx title="Autocomplete.jsx"
import { autocomplete } from '@algolia/autocomplete-js';
import React, { createElement, Fragment, useEffect, useRef } from 'react';
import { render } from 'react-dom';

export function Autocomplete(props) {
  const containerRef = useRef(null);

  useEffect(() => {
    if (!containerRef.current) {
      return undefined;
    }

    const search = autocomplete({
      container: containerRef.current,

>       placeholder: 'Search',

Reviewer: We can remove this option to be on par with the default Autocomplete JS options.

```suggestion
```

      renderer: { createElement, Fragment },
      render({ children }, root) {
        render(children, root);
      },
      ...props,
    });

    return () => {
      search.destroy();
    };
  }, [props]);

  return (
    <div
      ref={containerRef}
    />
  );
}
```

## Using the component

Now that you've created an `<Autocomplete/>` component, you can use it in your React application.

> The example component sets the [`placeholder`](autocomplete-js#placeholder) and [`container`](autocomplete-js/#container) options, but lets everything else get configured through props.

Reviewer: ```suggestion
The example component sets the [`container`](autocomplete-js/#container) option to mount Autocomplete on your component, but lets everything else get configured through props.
```

The example below sets [`openOnFocus`](autocomplete-js#openonfocus) and [sources](sources) (via [`plugins`](plugins)) through props. This example uses [recent searches](adding-recent-searches) as a [source](sources), but you could use anything else you want.

> The  Autocomplete library provides the [`createLocalStorageRecentSearchesPlugin`](createlocalstoragerecentsearchesplugin) function for creating a recent searches [plugin](plugins) out-of-the-box. To use it, you need to provide a `key` and `limit`.

Reviewer: I think it would make more sense not to use a plugin here, but rather to use `getSources`. This would be the opportunity for us to explain that the last step for Autocomplete to work with React is to provide `createElement` to all highlighting utils (because it defaults to `preact.createElement`.

```jsx
import { highlightHit } from '@algolia/autocomplete-js';
import React, { createElement } from 'react';

function AutocompleteItem({ hit }) {
  return (
    <a href={hit.url} className="aa-ItemLink">
      <div className="aa-ItemContent">
        <div className="aa-ItemTitle">
          {highlightHit({
            hit,
            attribute: 'name',
            createElement,
          })}
        </div>
      </div>
    </a>
  );
}
```

(This applies for `highlightHit`, `snippetHit`, ` reverseHighlightHit` and `reverseSnippetHit`.)

And finally, add a note that all officially supported plugins work with React.

The `key` can be any string and is required to differentiate search histories if you have multiple autocompletes on one page. The `limit` defines the maximum number of recent searches to display.

> ```js title=App.js"

Reviewer: ```suggestion
```jsx title=App.jsx"
```

import React from 'react';

> import { Autocomplete } from '/components/Autocomplete';

Reviewer: ```suggestion
import { Autocomplete } from './components/Autocomplete';
```

import { createLocalStorageRecentSearchesPlugin } from 'autocomplete-plugin-recent-searches';

const recentSearchesPlugin = createLocalStorageRecentSearchesPlugin({
  key: 'RECENT_SEARCH',
  limit: 5,
});

function App() {
  return (
    <div className="app-container">
      <h1>React Application</h1>
      <Autocomplete
        plugins={[recentSearchesPlugin]}
        openOnFocus={true}
      />
    </div>
  );
}

export default App;
```

> ## Further UI customization

Reviewer: Smooth ending 👍

Don't forget to [clean up the effect](https://reactjs.org/docs/hooks-reference.html#cleaning-up-an-effect) by returning a function that destroys the Autocomplete instance.

```jsx title="Autocomplete.jsx"
import { autocomplete } from '@algolia/autocomplete-js';
import React, { createElement, Fragment, useEffect, useRef } from 'react';
import { render } from 'react-dom';

export function Autocomplete(props) {
  const containerRef = useRef(null);

  useEffect(() => {
    if (!containerRef.current) {
      return undefined;
    }

    const search = autocomplete({
      container: containerRef.current,
      renderer: { createElement, Fragment },
      render({ children }, root) {
        render(children, root);
      },
      ...props,
    });

    return () => {
      search.destroy();
    };
  }, [props]);

  return (
    <div
      ref={containerRef}
    />
  );
}
```

## Using the component

> Now that you've created an `<Autocomplete/>` component, you can use it in your React application.

Reviewer: ```suggestion
Now that you've created an `<Autocomplete />` component, you can use it in your React application.
```

> As previously state, the example component sets only the [`container`](autocomplete-js/#container) option, but lets all other [options](basic-options) get configured through props.

Reviewer: ```suggestion
As previously stated, the example component sets only the [`container`](autocomplete-js/#container) option, but lets all other [options](basic-options) get configured through props.
```

> As previously state, the example component sets only the [`container`](autocomplete-js/#container) option, but lets all other [options](basic-options) get configured through props.

Reviewer: Is this paragraph needed as it was already said above?

Author: True -- https://github.com/algolia/autocomplete.js/pull/442/commits/8c6843808f31dcaa6878e49af294304501700df2

The usage below sets [`openOnFocus`](autocomplete-js#openonfocus) and [sources](sources) through props. This example uses an [Algolia index](https://www.algolia.com/doc/faq/basics/what-is-an-index/) as a [source](sources), but you could use anything else you want, including [plugins](plugins). For more information on using Algolia as a source, check out the [Getting Started guide](getting-started).

```jsx title=App.jsx"
import React, { createElement } from 'react';
import { Autocomplete } from './components/Autocomplete';

> import { ProductItem } from './components/ProductItem';

Reviewer: Can we move these two local imports after the package imports?

Author: https://github.com/algolia/autocomplete.js/pull/442/commits/78bd3c2913527485b9c91dec424011eb7ecee7d8

                      indexName: "instant_search",
                      query,
                    }
                  ]
                });
              },
              templates: {
                item({ item }) {
                  return <ProductItem hit={item} />;
                }
              }
            }
          ];
        }
      />
    </div>
  );
}

export default App;
```

### Creating templates

The example above passes `<ProductItem />`, another React component, for the `item` [template](templates). When creating templates, there's one thing to keep in mind. If you're using the highlighting and snippeting utilities, you must pass them React's `createElement` function. Without doing this, the utilities default to `preact.createElement` and won't work properly.

The highlighting and snippeting utilities are:
- [`highlightHit`](highlighthit)
- [`snippetHit`](snippethit)
- [`reverseHighlightHit`](reversehighlighthit)
- [`reverseSnippetHit`](reversesnippethit)

Here's an example using [`highlightHit`](highlighthit):

```jsx title="ProductItem.jsx"
import { highlightHit } from '@algolia/autocomplete-js';
import React, { createElement } from 'react';

> export function AutocompleteItem({ hit }) {

Reviewer: ```suggestion
export function ProductItem({ hit }) {
```
